### ISSUE TITLE
DATAprepSE function

### ISSUE THREAD

**Issue by elissonnog** [loubator/MultiRNAflow#2]
Hello. I m trying to use your package for test but I cant pass to the first function. There is some name dependency on it very unclear. I tried copy the pattern of your example and I still dont know what is wrong with the name of the samples. Can you help me?

 colnames(count_data) <- c("gene", "nt_t0_r1", "nt_t0_r2", "nt_t0_r3", 
+                           "w3_t3_r1", "w3_t3_r2", "w3_t3_r3", 
+                           "w8_t8_r1", "w8_t8_r2", "w8_t8_r3", 
+                           "w17_t17_r1", "w17_t17_r2", "w17_t17_r3", 
+                           "w27_t27_r1", "w27_t27_r2", "w27_t27_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3,
+                             colData = colData)
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.

**Comment by elissonnog** [comment]
Another try using only characters as individual name

colnames(count_data) <- c("gene", "nt_t0_r1", "nt_t0_r2", "nt_t0_r3", 
+                           "wt_t3_r1", "wt_t3_r2", "wt_t3_r3", 
+                           "wv_t8_r1", "wv_t8_r2", "wv_t8_r3", 
+                           "wg_t17_r1", "wg_t17_r2", "wg_t17_r3", 
+                           "ww_t27_r1", "ww_t27_r2", "ww_t27_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3)
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.

**Comment by elissonnog** [comment]
Using a mix

 colnames(count_data) <- c("gene", "nt_t0_r1", "nt_t0_r2", "nt_t0_r3", 
+                           "3_t3_r1", "3_t3_r2", "3_t3_r3", 
+                           "8_t8_r1", "8_t8_r2", "8_t8_r3", 
+                           "17_t17_r1", "17_t17_r2", "17_t17_r3", 
+                           "27_t27_r1", "27_t27_r2", "27_t27_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3)
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.

**Comment by elissonnog** [comment]
using the same name

colnames(count_data) <- c("gene", "sp_t0_r1", "sp_t0_r2", "sp_t0_r3", 
+                           "sp_t3_r1", "sp_t3_r2", "sp_t3_r3", 
+                           "sp_t8_r1", "sp_t8_r2", "sp_t8_r3", 
+                           "sp_t17_r1", "sp_t17_r2", "sp_t17_r3", 
+                           "sp_t27_r1", "sp_t27_r2", "sp_t27_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3)
Error in dimnames(x) <- dn : 
  length of 'dimnames' [1] not equal to array extent

**Comment by elissonnog** [comment]
colnames(count_data) <- c("gene", "x_nt_t0_r1", "c_nt_t0_r2", "v_nt_t0_r3", 
+                           "b_w3_t3_r1", "n_w3_t3_r2", "m_w3_t3_r3", 
+                           "a_w8_t8_r1", "s_w8_t8_r2", "d_w8_t8_r3", 
+                           "f_w17_t17_r1", "g_w17_t17_r2", "h_w17_t17_r3", 
+                           "j_w27_t27_r1", "k_w27_t27_r2", "l_w27_t27_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=2,
+                             Time.position=3,
+                             Individual.position=1,
+                             colData = colData)
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=2,
+                             Time.position=3,
+                             Individual.position=4,
+                             colData = colData)
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.

**Comment by loubator** [comment]
Hello, thank you for using my package.

I think I understand your problem.
"r1", "r2" ... represent the individual name (patient name, one mouse, one plant ...) so if you have "w3_t3_r1" and "w8_t3_r1" that means that the same individual "r1" belongs to two biological conditions which cause the error. You must replace for instance "w3_t3_r1", "w3_t3_r2", "w3_t3_r3" and "w8_t3_r1", "w8_t3_r2", "w8_t3_r3" by
- either "w3_t3_r1", "w3_t3_r2", "w3_t3_r3" and "w8_t3_r4", "w8_t3_r5", "w8_t3_r6"
- or "w3_t3_r1w3", "w3_t3_r2w3", "w3_t3_r3w3" and "w8_t3_r1w8", "w8_t3_r2w8", "w8_t3_r3w8".

As the error says, **every individual must have a unique name, must be associated to a unique group** and must be associated only once to each of the same 5 time measurements.

**Comment by elissonnog** [comment]
colnames(count_data) <- c("gene", "nt_t0_r1nt", "nt_t0_r2nt", "nt_t0_r3nt", 
+                           "w3_t3_r1w3", "w3_t3_r2w3", "w3_t3_r3w3", 
+                           "w8_t8_r1w8", "w8_t8_r2w8", "w8_t8_r3w8", 
+                           "w17_t17_r1w17", "w17_t17_r2w17", "w17_t17_r3w17", 
+                           "w27_t27_r1w27", "w27_t27_r2w27", "w27_t27_r3w27")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3)#,
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.
> # Extract sample information from column names
> colnames(count_data) <- c("gene", "nt_t0_r1", "nt_t0_r2", "nt_t0_r3", 
+                           "w3_t3_r4", "w3_t3_r5", "w3_t3_r6", 
+                           "w8_t8_r7", "w8_t8_r8", "w8_t8_r9", 
+                           "w17_t17_r10", "w17_t17_r11", "w17_t17_r12", 
+                           "w27_t27_r13", "w27_t27_r14", "w27_t27_r15")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=2,
+                             Individual.position=3)#,
Error in ColnamesToFactors(ExprData = dataColToFct, Column.gene = Column.gene,  : 
  Every individual must have a unique name, must be associated to a unique group and must be associated only once to each of the same 5 time measurements.

**Comment by loubator** [comment]
## Question

Wait a minute, do you mean that

_c("gene",
"nt_t0_r1nt", "nt_t0_r2nt", "nt_t0_r3nt",
"w3_t3_r1w3", "w3_t3_r2w3", "w3_t3_r3w3",
"w8_t8_r1w8", "w8_t8_r2w8", "w8_t8_r3w8",
"w17_t17_r1w17", "w17_t17_r2w17", "w17_t17_r3w17", 
"w27_t27_r1w27", "w27_t27_r2w27", "w27_t27_r3w27")_

represent all your colnames ? I thought it was a sample of your colnames.
Because, in that case, you have only one time point per biological condition (nt, w3, w8, w17, w27), which is problematic.

Can you explain your experimental design ?

I suggest two analysis.

## Analysis 1

If, as supposed above, you have only one time point per biological condition (nt, w3, w8, w17, w27) then you should keep the names above and write

```
resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
                             Column.gene=1,
                             Group.position=1,
                             Time.position=NULL,
                             Individual.position=3)
```

## Analysis 2

If what matters to you is the timecourse and you do not mind about (nt, w3, w8, w17, w27) then your colnames must be

_c("gene",
"t0_r1", "t0_r2", "t0_r3",
"t1_r1", "t1_r2", "t1_r3",
"t2_r1", "t2_r2", "t2_r3",
"t3_r1", "t3_r2", "t3_r3", 
"t4_r1", "t4_r2", "t4_r3")_

as explained in the manual and in the vignette, and you must write

```
resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
                             Column.gene=1,
                             Group.position=NULL,
                             Time.position=1,
                             Individual.position=2)
```

## Important

If you want to realize an analysis with time and biological condition, you must have several (and identical) times point for all your biological conditions.

**Comment by elissonnog** [comment]
I have 3 samples measured during 5 times. 
Nt 1, 2 ,3 (time 0 - no treatment)
w3 1 2 3
w8 1 2 3
w17 1 2 3
w27 1 2 3

**Comment by elissonnog** [comment]
colnames(count_data) <- c("gene", "nt_t0_r1nt", "nt_t0_r2nt", "nt_t0_r3nt", 
+                           "w3_t3_r1w3", "w3_t3_r2w3", "w3_t3_r3w3", 
+                           "w8_t8_r1w8", "w8_t8_r2w8", "w8_t8_r3w8", 
+                           "w17_t17_r1w17", "w17_t17_r2w17", "w17_t17_r3w17", 
+                           "w27_t27_r1w27", "w27_t27_r2w27", "w27_t27_r3w27")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                             Column.gene=1,
+                             Group.position=1,
+                             Time.position=NULL,
+                             Individual.position=3)#,
Error in dimnames(x) <- dn : 
  length of 'dimnames' [1] not equal to array extent
  
  
  
  colnames(count_data) <-c("gene",
+ "t0_r1", "t0_r2", "t0_r3",
+ "t1_r1", "t1_r2", "t1_r3",
+ "t2_r1", "t2_r2", "t2_r3",
+ "t3_r1", "t3_r2", "t3_r3",
+ "t4_r1", "t4_r2", "t4_r3")
> resDATAprepSE0 <- DATAprepSE(RawCounts=count_data,
+                              Column.gene=1,
+                              Group.position=NULL,
+                              Time.position=1,
+                              Individual.position=2)
Error in dimnames(x) <- dn : 
  length of 'dimnames' [1] not equal to array extent

**Comment by loubator** [comment]
I simulated a raw count data (code below), I used the same column names and it works perfectly fine.

```
## simulation of a raw count data
count_matrix <- matrix(sample(seq_len(1000), size=1500, replace=TRUE), nrow=100)
count_data <- data.frame(Gene=paste0("gene", seq_len(100)), count_matrix)
colnames(count_data) <-c("gene",
                         "t0_r1", "t0_r2", "t0_r3",
                         "t1_r1", "t1_r2", "t1_r3",
                         "t2_r1", "t2_r2", "t2_r3",
                         "t3_r1", "t3_r2", "t3_r3",
                         "t4_r1", "t4_r2", "t4_r3")

## for a quick view of the data
str(count_data) 

## use of DATAprepSE function
resDATAprepSE0 <- MultiRNAflow::DATAprepSE(RawCounts=count_data,
                                           Column.gene=1,
                                           Group.position=NULL,
                                           Time.position=1,
                                           Individual.position=2)
```

Are you sure that your `count_data` has gene names in the first column (not set as row names) ?
If you have set gene names as row names, set `Column.gene=NULL`, however I recommand to have a column of your data with gene names.

**Comment by elissonnog** [comment]
I was expecting to use the colnames in a different way based on my design as I mentioned before, I have replicates and times that I want to be considered. But I am trying to reproduce even with the example you posted and its not working

 colnames(count_data) <-c("gene",
+                          "t0_r1", "t0_r2", "t0_r3",
+                          "t1_r1", "t1_r2", "t1_r3",
+                          "t2_r1", "t2_r2", "t2_r3",
+                          "t3_r1", "t3_r2", "t3_r3",
+                          "t4_r1", "t4_r2", "t4_r3")
> 
> ## for a quick view of the data
> str(count_data) 
tibble [14,446 × 16] (S3: tbl_df/tbl/data.frame)
 $ gene : chr [1:14446] "ENSG00000000003" "ENSG00000000419" "ENSG00000000457" "ENSG00000000460" ...
 $ t0_r1: int [1:14446] 1435 2405 515 2017 1 230 5276 1101 1321 491 ...
 $ t0_r2: int [1:14446] 1268 2308 540 1939 3 161 5317 1080 1204 446 ...
 $ t0_r3: int [1:14446] 1788 2801 607 2093 6 424 5847 1568 1134 585 ...
 $ t1_r1: int [1:14446] 1315 2538 585 1910 0 198 5606 1386 1329 426 ...
 $ t1_r2: int [1:14446] 1439 2653 535 1929 0 265 6263 1515 1375 566 ...
 $ t1_r3: int [1:14446] 1699 2746 587 2036 1 588 6457 1403 1235 511 ...
 $ t2_r1: int [1:14446] 1951 3036 666 2417 0 279 8061 1278 1609 638 ...
 $ t2_r2: int [1:14446] 2057 3324 688 2327 2 334 7834 1426 1677 553 ...
 $ t2_r3: int [1:14446] 1643 2818 505 1974 5 285 7056 1480 1269 543 ...
 $ t3_r1: int [1:14446] 1394 2370 468 1897 1 114 6317 1179 1595 576 ...
 $ t3_r2: int [1:14446] 1533 2379 477 1897 2 224 7711 1683 1576 582 ...
 $ t3_r3: int [1:14446] 1578 2936 423 1993 0 129 7116 1948 1314 621 ...
 $ t4_r1: int [1:14446] 1704 2219 428 1821 0 282 7415 1626 1469 420 ...
 $ t4_r2: int [1:14446] 1721 2723 392 1865 1 219 7991 1740 1542 589 ...
 $ t4_r3: int [1:14446] 2471 2832 519 1912 0 357 8516 1869 1592 583 ...
> 
> ## use of DATAprepSE function
> resDATAprepSE0 <- MultiRNAflow::DATAprepSE(RawCounts=count_data,
+                                            Column.gene=1,
+                                            Group.position=NULL,
+                                            Time.position=1,
+                                            Individual.position=2)
Error in dimnames(x) <- dn : 
  length of 'dimnames' [1] not equal to array extent

**Comment by loubator** [comment]
For the example I posted, I checked again and it works perfectly fine on my computer so I have no idea what is going on.
Have you try to just copy, paste and execute on your console the code without any modification ?

However, for your dataset, there is something odd. I think the problem is the class of your dataset you set.
```
str(count_data)
tibble [14,446 × 16] (S3: tbl_df/tbl/data.frame)
```
The class of your `count_data` is `tibble` and for me
```
str(count_data) 
'data.frame':	100 obs. of  16 variables:
```
it is `data.frame`. I suggest to use the command
```
count_data <- as.data.frame(count_data)
```
before using the function `DATAprepSE()` because `count_data` must be a `data.frame`.
